### DISCUSSION TITLE
How to deactivate inactive http input sources in a switch?

### DISCUSSION THREAD

**Discussion by liquidpeter** [savonet/liquidsoap#2961] [User Support]
I have a fixed weekly schedule where I want to switch between 
between the streams of our local radio stations.
For example: Monday plays station A with its URL, Tuesday plays stationB with its URL and so on.

So I use a "switch" between several http input sources.
In total there are about 10 http input stream sources in the switch.
Unfortunately they all get fetched even if their time predicate is false.
I tried using "server.execute" to stop stations that are not being used at the moment.
But this has no effect: The http source status remains "polling".

Do you have any idea how I can deactivate (stop polling) http sources in a switch (and start them later again)?

Background:
I used to use only 2 streams and set the URL dynamically using an external program to change the stream URLs. 
But now I am wondering if there is a simple way to use the switch with fixed time slots to 
to select the source depending on the time predicate.
I tried to enable and disable switch sources using thread.run.

**Comment by smimram** [comment]
Quick side note: maybe would we like to introduce a variant of `switch` which does not animate sources which are not used?

**Comment by toots** [comment]
Switch do not animate unused sources 🙂

The problem here is that `input.http` is an _active source_ so it is animated all the time, regardless of its use. The reason being that, when we ask the source to be ready is needs to be_immediately_ ready so we need to keep an active connection.

@liquidpeter the sources returned by `input.http` all have a `start` and `stop` method. I would suggest to use those to manually turn them on and off when you do or don't need them.

**Comment by liquidpeter** [comment]
Thanks for your answers!, 

Calling "start" and "stop" is exactly what I tried when I said I used the "server.execute"  commands.
But: If I call "stop", I see the URL status keeps "polling".
If the source URL is not available (for example invalid URL is set) I see continously error output due to failing connect.
(I assume this means the process tries to reconnect)

Maybe its easier if you see my code. This is it:
```
#input-stream-http
radio1 = fallback(id="radio1", track_sensitive=false, [
    input.http(id="radio1-url1", "http://radio1"),
    input.http(id="radio1-url2", "http://radio2"),
    single(id="radio1-fallback", "/media/radio2-fallback.mp3")
])

radio2 = fallback(id="radio2", track_sensitive=false, [
    input.http(id="radio2-url1", "http://radio1"),
    input.http(id="radio2-url2", "http://radio2"),
    single(id="radio2-fallback", "/media/radio2-fallback.mp3")
])

# go on and define radio3 to radio7 …

schedule = switch(id="schedule", track_sensitive=false, all_predicates=true, [
        ({1w06h-2w06h}, radio1),
        ({2w06h-3w06h}, radio2),
        ({3w06h-4w06h}, radio3),
        ({4w06h-5w06h}, radio4),
        ({5w06h-6w06h}, radio5),
        ({6w06h-7w06h}, radio6),
        ({7w06h-1w06h}, radio7),
])

output.icecast(
    %vorbis.cbr(samplerate=44100, channels=2, bitrate=256),
    host="localhost", mount="radio", password="secret", port=8000, user="source",
    mksafe(buffer(radio))
)

def start_station(name)
    ignore(server.execute("#{name}-url1.start"))
    ignore(server.execute("#{name}-url2.start"))
end

def stop_station(name)
    ignore(server.execute("#{name}-url1.stop"))
    ignore(server.execute("#{name}-url2.stop"))
end

# activate sources (an hour earlier)
def activate_sources()
    if (1w05h-2w06h) then
        start_station("radio1")
    end
    if (2w05h-3w06h) then
        start_station("radio2")
    end
    # more conditions to start depending on weekday
end

# stop unused sources
def deactivate_sources()
    if (1w06h-2w06h) then 
        stop_station("radio2")
        stop_station("radio3")
        stop_station("radio4")
        stop_station("radio5")
        stop_station("radio6")
        stop_station("radio7")
    end
    if (2w06h-3w06h) then 
        stop_station("radio1")
        stop_station("radio3")
        stop_station("radio4")
        stop_station("radio5")
        stop_station("radio6")
        stop_station("radio7")
    end
    # more conditions to stop depending on weekday
end

thread.run(every=60., activate_sources)
thread.run(every=60., deactivate_sources)
```

I also tries to use the telnet interface to "stop" the sources with the same result.

I use Liquidsoap 2.1.4 (from opam) on Ubuntu 22.04.

Thanks and BR, Peter

update: I added some missing "ignore(…)" statements

**Reply by liquidpeter** [reply]
I think, I found the main issue: The "stop" does not stop polling an unavailable source.
The first URL I used in the switch is not available at all. 
If one tries to connect to this URL using curl, there process hangs because it waits for an response.
If I try to stop this input source with the "stop" command, the result is "Done!", but the URL status is "polling" afterwards.
In the logs I see output every some seconds like
`[tcp @ 0x7faa7c06a400] Connection to tcp://<domain>:8000 failed: Connection timed out`
So this is issue is independent of the switch command.

**Reply by toots** [reply]
Okay I see. Is there anything that you think needs to be fixed on our end?

**Reply by liquidpeter** [reply]
It would be great if the "stop" command really stops the input trying to periodically access the unavailable stream.
To me this sounds like a bug, but it is not really annoying.

But basically the switch and everything else works as expected.
PS: I just found out how to write out the current status of all sources to an json file 
and how to write peak and rms into an log file to externally plot the current stream.
It's fantastic that I can do all these things in liquidsoap now.
There have been many improvements here in recent years. 
Thank you for your great work!

**Comment by Moonbase59** [comment]
Nowadays, there’s more syndication going on than probably anticipated. I think—in the long run—we should find a way of syndicating remote streams without keeping open a running stream 24/7 for a show that plays 2 hours a week. I noticed this when agreeing to "take over" a few 2-hour shows from a friend’s radio and ended up blocking 10 of their listener slots all the time (and they only got 50).
